### DISCUSSION TITLE
How to replace a keycloak builtin EventStoreProvider?

### DISCUSSION THREAD

**Discussion by danielFesenmeyer** [keycloak/keycloak#13837] [Q&A]
Hi,

we have implemented our own "Modified"EventStoreProvider based on JpaEventStoreProvider.
(We did this, because we want to use the functionality provided by this PR: https://github.com/keycloak/keycloak/pull/8822.)

Our "Modified"EventStoreProviderFactory had the ID "jpa-modified" and a higher order than the builtin JpaEventStoreProviderFactory with ID "jpa".
This worked fine with keycloak-legacy version 18: the "Modified"EventStoreProvider was loaded instead of JpaEventStoreProvider.

But it worked no longer with Keycloak.X 19: JpaEventStoreProvider was loaded instead of "Modified"EventStoreProvider.

I tried to fix that by explicitly specifying "jpa-modified" as provider, with the env var mechanism:
`KC_SPI_EVENTS_STORE_PROVIDER=jpa-modified`

But the latter did not work, Keycloak did not recognize the setting - the used provider was still "jpa".
It seems that there is a problem with handling SPI names in camel-case format (_eventsStore_, in this case).
Hence, I tried the CLI arg mechanism:
`--spi-events-store-provider=jpa-modified`

Now, surprisingly the "map" event store was used. After some debugging I found out that this is due to the following implementation in StoragePropertyMappers:
https://github.com/keycloak/keycloak/blob/0f91beeca43be5b5f171183223d89e0451da9f20/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/StoragePropertyMappers.java#L318
When no storage config is found, "jpa" is used, otherwise "map". The specified value is completely ignored.
I would expect an error or  at least a warning that I have specified an unsupported option.
Furthermore, it seems that it is currently not possible to use any custom implementation, for example an implementation based on a timeseries db.

What finally worked in our use case: I've changed the ID of our "Modified"EventStoreProviderFactory to "jpa". This works. because the ServiceLoader mechanism loads our class before the "overwritten" Keycloak class:
https://github.com/keycloak/keycloak/blob/0f91beeca43be5b5f171183223d89e0451da9f20/services/src/main/java/org/keycloak/provider/ProviderManager.java#L94

But from my point of view, the mechanism would be more reliable, if the order of the ProviderFactories would be considered, in case there is more than one matching.

I'm just wondering if someone has already solved a similar issue and maybe knows a better solution?

**Comment by fperot74** [comment]
Hi,

As discussed in the keycloak dev mailing list, we had the same problem for our custom user provider.
It was working correctly with KC18 with this configuration in conf/keycloak.conf:
spi-user-jpa-enabled=false
spi-user-our-jpa-enabled=true

When migrating to KC20, it was not working anymore. Our custom user provider is not seen by Keycloak (not displayed in Server info/Providers) and a message tells at startup that the "jpa" provider could not be found.
Renaming our-jpa to jpa made it work.
